File: kangaroo_app_sdk_ios/ios/Classes/features/getCrmFields/GetCrmFieldsHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class GetCrmFieldsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_crm_fields"

    var eventChannel: String = "customer_sdk/events/get_crm_fields"

    func onMethodCall(call: FlutterMethodCall) -> Void? {
        GetCrmFieldsHandler.getCrmFields(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getCrmFields(call: FlutterMethodCall) {
        GetCrmFieldsApi().getCrmFields()

        

        

        
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        GetCrmFieldsApi().getCrmFieldsState
            .serializeGetCrmFieldsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}

File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/getCrmFields/GetCrmFieldsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.getCrmFields

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.getCrmFields.GetCrmFieldsApi
import features.getCrmFields.models.CrmFields

import features.getCrmFields.serializeGetCrmFieldsState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class GetCrmFieldsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_crm_fields"

    override val eventChannel: String
        get() = "customer_sdk/events/get_crm_fields"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getCrmFields(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getCrmFields(call: MethodCall): String? {
            val result = GetCrmFieldsApi().getCrmFields().toJsonResult<CrmFields>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        GetCrmFieldsApi().getCrmFieldsState.serializeGetCrmFieldsState().observe { result ->
            pushSerializedResultToEventSink<GetCrmFieldsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
